File: framework/src/play/src/main/scala/play/api/ConfiguredLogging.scala
/*
 * Copyright (C) 2009-2018 Lightbend Inc. <https://www.lightbend.com>
 */

package play.api

import org.slf4j.ILoggerFactory

trait ConfiguredLogging {

  val environment: Environment

  /**
   * Configures the SLF4J logger factory.  This is where LoggerConfigurator is
   * called from.
   *
   * @param configuration play.api.Configuration
   * @return the app wide ILoggerFactory.  Useful for testing and DI.
   */
  def configureLoggerFactory(configuration: Configuration): ILoggerFactory = {
    val loggerFactory: ILoggerFactory = LoggerConfigurator(environment.classLoader).map { lc =>
      lc.configure(environment, configuration, Map.empty)
      lc.loggerFactory
    }.getOrElse(org.slf4j.LoggerFactory.getILoggerFactory)

    if (shouldDisplayLoggerDeprecationMessage(configuration)) {
      val logger = loggerFactory.getLogger("application")
      logger.warn("Logger configuration in conf files is deprecated and has no effect. Use a logback configuration file instead.")
    }

    loggerFactory
  }

  /**
   * Checks if the path contains the logger path
   * and whether or not one of the keys contains a deprecated value
   *
   * @param appConfiguration The app configuration
   * @return Returns true if one of the keys contains a deprecated value, otherwise false
   */
  def shouldDisplayLoggerDeprecationMessage(appConfiguration: Configuration): Boolean = {
    import scala.collection.JavaConverters._
    import scala.collection.mutable

    val deprecatedValues = List("DEBUG", "WARN", "ERROR", "INFO", "TRACE", "OFF")

    // Recursively checks each key to see if it contains a deprecated value
    def hasDeprecatedValue(values: mutable.Map[String, AnyRef]): Boolean = {
      values.exists {
        case (_, value: String) if deprecatedValues.contains(value) =>
          true
        case (_, value: java.util.Map[_, _]) =>
          val v = value.asInstanceOf[java.util.Map[String, AnyRef]]
          hasDeprecatedValue(v.asScala)
        case _ =>
          false
      }
    }

    if (appConfiguration.underlying.hasPath("logger")) {
      appConfiguration.underlying.getAnyRef("logger") match {
        case value: String =>
          hasDeprecatedValue(mutable.Map("logger" -> value))
        case value: java.util.Map[_, _] =>
          val v = value.asInstanceOf[java.util.Map[String, AnyRef]]
          hasDeprecatedValue(v.asScala)
        case _ =>
          false
      }
    } else {
      false
    }
  }
}

File: framework/src/play/src/main/java/play/ConfiguredLogging.java
/*
 * Copyright (C) 2009-2018 Lightbend Inc. <https://www.lightbend.com>
 */

package play;

import com.typesafe.config.Config;
import org.slf4j.ILoggerFactory;

import java.util.Arrays;
import java.util.Collection;
import java.util.Map;

public interface ConfiguredLogging
{
    Environment environment();

    default ILoggerFactory configureLoggerFactory(final Config configuration) {
        final Environment environment = environment();
        final ILoggerFactory loggerFactory = LoggerConfigurator.apply(environment.classLoader()).map(lc -> {
            lc.configure(environment, configuration);
            return lc.loggerFactory();
        }).orElseGet(org.slf4j.LoggerFactory::getILoggerFactory);

        if (shouldDisplayLoggerDeprecationMessage(configuration)) {
            loggerFactory
                    .getLogger("application")
                    .warn("Logger configuration in conf files is deprecated and has no effect. Use a logback configuration file instead.");
        }

        return loggerFactory;
    }

    default boolean shouldDisplayLoggerDeprecationMessage(final Config configuration) {
        return configuration.hasPath("logger") && isDeprecated(
                configuration.getAnyRef("logger"),
                Arrays.asList("DEBUG", "WARN", "ERROR", "INFO", "TRACE", "OFF")
        );
    }

    default boolean hasDeprecatedValue(final Map<String, Object> config, final Collection<String> deprecatedValues) {
        return config.values().stream().map(value -> isDeprecated(value, deprecatedValues)).findAny().orElse(false);
    }

    @SuppressWarnings("unchecked")
    default Boolean isDeprecated(Object value, Collection<String> deprecatedValues) {
        if (value instanceof String) {
            return deprecatedValues.contains(value);
        } else if (value instanceof Map) {
            return hasDeprecatedValue((Map<String, Object>) value, deprecatedValues);
        } else {
            return false;
        }
    }
}
